portindex2json/portindex2json.tcl

```diff
@@ -64,31 +64,82 @@ proc remove_extra_braces {text} {
     return [regsub -all {[\{\}]} $text ""]
 }
 
-set fd [open [lindex $argv 0] r]
-chan configure $fd -encoding utf-8
-while {[gets $fd line] >= 0} {
-    if {[llength $line] != 2} {
-        continue
-    }
-    set len [lindex $line 1]
-    set line [read $fd $len]
-    array unset portinfo
-    array set portinfo $line
-    array unset json_portinfo
-    foreach key [array names portinfo] {
-        if {$key == "categories" || $key == "variants" || [string match "depends_*" $key]} {
-            set json_portinfo($key) [tcl2json_list $portinfo($key)]
-        } elseif {$key == "maintainers"} {
-            set json_portinfo($key) [parse_maintainers $portinfo($key)]
-            set json_portinfo(closedmaintainer) [is_closedmaintainer $portinfo($key)]
-        } elseif {$key == "long_description"} {
-            set json_portinfo($key) [::json::write string [remove_extra_braces $portinfo($key)]]
-        } else {
-            set json_portinfo($key) [::json::write string $portinfo($key)]
+set filename 0
+
+for {set i 0} {$i < $argc} {incr i} {
+    set arg [lindex $argv $i]
+    switch -regex -- $arg {
+        {^-.+} {
+            if {$arg eq "-info"} {
+                incr i
+                set argument [split [lindex $argv $i] =]
+                if {[lindex $argument 1] eq ""} {
+                    puts stderr "\nError: No value provided for the key: [lindex $argument 0]. The correct syntaxt is -info key=value. Use -help for more\n"
+                    exit 1
+                } else {
+                    set json_info([lindex $argument 0]) [::json::write string [lindex $argument 1]]
+                }
+            } elseif {$arg eq "-help" || $arg eq "-h" || $arg eq "--help" } {
+                puts stdout "\nUsage: /path/to/tclsh portindex2json.tcl <path-to-portindex> -info key=vlaue"
+                puts stdout "\nEach '-info key=value' combinaton adds one key-value pair to the JSON output inside the parent key: \"info\""
+                puts stdout "\nYou may provide any number of '-info key=value' combinations."
+                puts stdout "\nThe arguments <path-to-portindex> and `-info key=value` are position independent.\n"
+                exit 0
+            } else {
+                puts stderr "\nError: $arg is an invalid argument. Expecting -info, -help.\n"
+                exit 1
+            }
+        }
+        default {
+            incr filename
+            if { [catch {open [lindex $argv $i] r} fid] } {
+                puts stderr "\nError: Could not open [lindex $argv $i]. Make sure the path to the PortIndex file is correct\n"
+                exit 1
+            } else {
+                set fd [open [lindex $argv $i] r]
+            }
         }
     }
-    lappend objects [::json::write object {*}[array get json_portinfo]]
 }
 
-chan configure stdout -encoding utf-8
-puts [::json::write array {*}$objects]
+if {[info exists fd] && $filename eq 1} {
```

Comment:
I would instead do "if fd doesn't exist, exit immediately" and do the rest of the code without a deeply nested if clause. For more that one file, see the other comment (you should probably not even get to this point).